File: git_visual.cs
// git_visual.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LibGit2Sharp;

class GitVisual
{
    private string repoPath;
    private bool graph;
    private bool all;
    private bool decorate;
    private bool oneline;
    private string author;
    private string since;
    private string until;
    private int maxCount;
    private bool color;
    private Repository repo;

    public GitVisual(string path, bool graph, bool all, bool decorate, bool oneline,
                     string author, string since, string until, int maxCount, bool color)
    {
        this.repoPath = path;
        this.graph = graph;
        this.all = all;
        this.decorate = decorate;
        this.oneline = oneline;
        this.author = author;
        this.since = since;
        this.until = until;
        this.maxCount = maxCount;
        this.color = color;
    }

    public void OpenRepo()
    {
        repo = new Repository(repoPath);
    }

    private string Colorize(string text, string colorCode)
    {
        if (!color) return text;
        var codes = new Dictionary<string, string>
        {
            {"red", "\x1b[91m"},
            {"green", "\x1b[92m"},
            {"yellow", "\x1b[93m"},
            {"blue", "\x1b[94m"},
            {"cyan", "\x1b[96m"},
            {"reset", "\x1b[0m"}
        };
        return codes[colorCode] + text + codes["reset"];
    }

    private IEnumerable<Commit> GetCommits()
    {
        var filter = new CommitFilter
        {
            SortBy = CommitSortStrategies.Time
        };
        if (all)
            filter.IncludeReachableFrom = repo.Refs;
        else
            filter.IncludeReachableFrom = repo.Head;
        if (!string.IsNullOrEmpty(author))
            filter.Author = author;
        // since/until можно добавить через фильтрацию после
        var commits = repo.Commits.QueryBy(filter);
        if (maxCount > 0)
            commits = commits.Take(maxCount);
        return commits;
    }

    private Dictionary<string, List<string>> GetRefs()
    {
        var dict = new Dictionary<string, List<string>>();
        foreach (var branch in repo.Branches)
        {
            if (branch.Tip != null)
            {
                string hash = branch.Tip.Sha.Substring(0, 7);
                if (!dict.ContainsKey(hash)) dict[hash] = new List<string>();
                dict[hash].Add(branch.FriendlyName);
            }
        }
        foreach (var tag in repo.Tags)
        {
            var target = tag.Target as Commit;
            if (target != null)
            {
                string hash = target.Sha.Substring(0, 7);
                if (!dict.ContainsKey(hash)) dict[hash] = new List<string>();
                dict[hash].Add(tag.FriendlyName);
            }
        }
        // HEAD
        if (repo.Head != null && repo.Head.Tip != null)
        {
            string hash = repo.Head.Tip.Sha.Substring(0, 7);
            if (!dict.ContainsKey(hash)) dict[hash] = new List<string>();
            dict[hash].Add("HEAD");
        }
        return dict;
    }

    public void ShowGraph()
    {
        var commits = GetCommits();
        var refs = GetRefs();
        int count = commits.Count();
        int i = 0;
        foreach (var c in commits)
        {
            bool last = i == count - 1;
            string connector = last ? "└── " : "├── ";
            string hashShort = c.Sha.Substring(0, 7);
            string refStr = "";
            if (decorate && refs.ContainsKey(hashShort))
                refStr = " (" + string.Join(", ", refs[hashShort]) + ")";
            if (oneline)
            {
                Console.WriteLine($"{connector}{Colorize(hashShort, "yellow")}{Colorize(refStr, "blue")} {Colorize(c.MessageShort, "reset")}");
            }
            else
            {
                Console.WriteLine($"{connector}{Colorize(hashShort, "yellow")}{Colorize(refStr, "blue")}");
                Console.WriteLine($"    {Colorize("Author:", "cyan")} {c.Author.Name} <{c.Author.Email}>");
                Console.WriteLine($"    {Colorize("Date:", "green")} {c.Author.When}");
                Console.WriteLine($"    {Colorize(c.MessageShort, "reset")}\n");
            }
            i++;
        }
    }

    static void Main(string[] args)
    {
        var path = ".";
        var graph = true;
        var all = false;
        var decorate = true;
        var oneline = false;
        string author = null;
        string since = null;
        string until = null;
        int maxCount = 0;
        bool color = true;

        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "--path": path = args[++i]; break;
                case "--graph": graph = bool.Parse(args[++i]); break;
                case "--all": all = true; break;
                case "--decorate": decorate = bool.Parse(args[++i]); break;
                case "--oneline": oneline = true; break;
                case "--author": author = args[++i]; break;
                case "--since": since = args[++i]; break;
                case "--until": until = args[++i]; break;
                case "-n": maxCount = int.Parse(args[++i]); break;
                case "--color": color = bool.Parse(args[++i]); break;
                case "--help": Console.WriteLine("Usage: ..."); return;
            }
        }

        var visual = new GitVisual(path, graph, all, decorate, oneline, author, since, until, maxCount, color);
        try
        {
            visual.OpenRepo();
            visual.ShowGraph();
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error: {ex.Message}");
            Environment.Exit(1);
        }
    }
}

File: git_visual.js
// git_visual.js
#!/usr/bin/env node
const simpleGit = require('simple-git');
const { program } = require('commander');
const chalk = require('chalk');

program
    .option('--graph', 'Show graph', true)
    .option('--all', 'Show all branches', false)
    .option('--decorate', 'Show refs', true)
    .option('--oneline', 'Compact output', false)
    .option('--author <name>', 'Filter by author')
    .option('--since <date>', 'Commits after date')
    .option('--until <date>', 'Commits before date')
    .option('-n, --max-count <number>', 'Limit commits', parseInt)
    .option('--color', 'Color output', true)
    .option('--path <path>', 'Repository path', '.')
    .parse(process.argv);

const opts = program.opts();
const git = simpleGit(opts.path);

async function getCommits() {
    const logOpts = {};
    if (opts.all) logOpts.all = true;
    if (opts.author) logOpts.author = opts.author;
    if (opts.since) logOpts.since = opts.since;
    if (opts.until) logOpts.until = opts.until;
    if (opts.maxCount) logOpts['-n'] = opts.maxCount;
    logOpts['--pretty'] = 'format:%h|%an|%ae|%ad|%s';
    const log = await git.log(logOpts);
    return log.all;
}

async function getRefs() {
    // Получаем ветки и теги
    const branches = await git.branch();
    const tags = await git.tags();
    const refs = {};
    for (const [name, branch] of Object.entries(branches.branches)) {
        const hash = branch.commit.substring(0, 7);
        if (!refs[hash]) refs[hash] = [];
        refs[hash].push(name);
    }
    for (const tag of tags.all) {
        // Упрощённо: получаем хеш для тега через show
        try {
            const show = await git.show(['-s', '--format=%H', tag]);
            const hash = show.substring(0, 7);
            if (!refs[hash]) refs[hash] = [];
            refs[hash].push(tag);
        } catch (e) {}
    }
    // HEAD
    const currentBranch = await git.branchLocal();
    const head = currentBranch.current;
    if (head) {
        const hash = currentBranch.branches[head].commit.substring(0, 7);
        if (!refs[hash]) refs[hash] = [];
        refs[hash].push('HEAD -> ' + head);
    }
    return refs;
}

function colorize(text, color) {
    if (!opts.color) return text;
    return chalk[color](text);
}

function printCommit(commit, prefix, last, refs) {
    const connector = last ? '└── ' : '├── ';
    const hashShort = commit.hash.substring(0, 7);
    const refStr = refs[hashShort] ? ' (' + refs[hashShort].join(', ') + ')' : '';
    if (opts.oneline) {
        console.log(`${prefix}${connector}${colorize(hashShort, 'yellow')}${colorize(refStr, 'blue')} ${colorize(commit.message, 'reset')}`);
    } else {
        console.log(`${prefix}${connector}${colorize(hashShort, 'yellow')}${colorize(refStr, 'blue')}`);
        console.log(`${prefix}    ${colorize('Author:', 'cyan')} ${commit.author_name} <${commit.author_email}>`);
        console.log(`${prefix}    ${colorize('Date:', 'green')} ${commit.date}`);
        console.log(`${prefix}    ${colorize(commit.message, 'reset')}\n`);
    }
}

async function main() {
    try {
        const commits = await getCommits();
        if (!commits.length) {
            console.log('No commits found.');
            return;
        }
        const refs = await getRefs();
        for (let i = 0; i < commits.length; i++) {
            const last = i === commits.length - 1;
            printCommit(commits[i], '', last, refs);
        }
    } catch (err) {
        console.error('Error:', err.message);
        process.exit(1);
    }
}

main();
